knative/deployer.go

```diff
@@ -7,73 +7,89 @@ import (
 	"os"
 	"strings"
 
-	"github.com/boson-project/faas"
-	"github.com/boson-project/faas/k8s"
 	commands "knative.dev/client/pkg/kn/commands"
 	"knative.dev/client/pkg/kn/core"
+
+	"github.com/boson-project/faas"
+	"github.com/boson-project/faas/k8s"
 )
 
 // TODO: Use knative.dev/serving/pkg/client/clientset/versioned/typed/serving/v1
 // NewForConfig gives you the client, and then you can do
 // client.Services("ns").Get("name")
 
-func NewDeployer() *Deployer {
-	return &Deployer{Namespace: faas.DefaultNamespace}
-}
-
 type Deployer struct {
+	// Namespace with which to override that set on the default configuration (such as the ~/.kube/config).
+	// If left blank, deployment will commence to the configured namespace.
 	Namespace string
-	Verbose   bool
+	// Verbose logging enablement flag.
+	Verbose bool
 }
 
-func (deployer *Deployer) Deploy(name, image string) (address string, err error) {
+func NewDeployer() *Deployer {
+	return &Deployer{}
+}
 
-	project, err := k8s.ToSubdomain(name)
+func (d *Deployer) Deploy(f faas.Function) (err error) {
+	// k8s does not support service names with dots.  so encode it such that
+	// www.my-domain,com -> www-my--domain-com
+	encodedName, err := k8s.ToSubdomain(f.Name)
 	if err != nil {
 		return
 	}
-	nn := strings.Split(name, ".")
-	if len(nn) < 3 {
-		err = fmt.Errorf("invalid service name '%v', must be at least three parts.\n", name)
-		return
-	}
-
-	subDomain := nn[0]
-	domain := strings.Join(nn[1:], ".")
 
+	// Capture output in a buffer if verbose is not enabled for output on error.
 	var output io.Writer
-	if deployer.Verbose {
+	if d.Verbose {
 		output = os.Stdout
 	} else {
 		output = &bytes.Buffer{}
 	}
 
+	// FIXME(lkinglan): The labels set explicitly here may interfere with the
+	// cluster configuraiton steps described in the documentation, and may also
+	// break on multi-level subdomains or if they are out of sync with that
+	// configuration.  These could be removed from here, and instead the cluster
+	// expeted to be configured correctly.  It is a future enhancement that an
+	// attempt to deploy a publicly accessible Function of a hithertoo unseen
+	// TLD+1 will modify this config-map.
+	// See https://github.com/boson-project/faas/issues/47
+	nn := strings.Split(f.Name, ".")
+	if len(nn) < 3 {
+		err = fmt.Errorf("invalid service name '%v', must be at least three parts.\n", f.Name)
+		return
+	}
+	subDomain := nn[0]
+	domain := strings.Join(nn[1:], ".")
```

Comment:
Create an issue to discuss whether we should expand this to a `publicsuffix`-package-based implementation, omit entirely relying on proper cluster configuration, or some hybrid.

Reply:
Encapsulated in #85 and #86, and tangentially with #47